src/child.rs

```diff
@@ -35,11 +57,42 @@ use tokio::process;
 /// ```
 #[derive(Debug)]
 pub struct RemoteChild<'s> {
-    pub(crate) session: &'s Session,
-    pub(crate) channel: Option<process::Child>,
+    session: &'s Session,
+    imp: RemoteChildImp,
+
+    stdin: Option<ChildStdin>,
+    stdout: Option<ChildStdout>,
+    stderr: Option<ChildStderr>,
 }
 
 impl<'s> RemoteChild<'s> {
+    pub(crate) fn new(session: &'s Session, mut imp: RemoteChildImp) -> Self {
+        Self {
+            session,
+
+            stdin: match &mut imp {
```

Comment:
Thanks for the macro.

I tried to come up with a macro before, but I failed.

Reply:
I also think of an alternative way using trait, but without any allocation:

```
enum RemoteChildImp {
    ProcessImpl(process_impl::Child),

    #[cfg(feature = "mux_client")]          
    MuxClientImpl(mux_client_impl::Child),
}

impl RemoteChildImp {
    fn to_trait_obj(&self) -> &dyn RemoteChildTrait {
        match self {
            RemoteChildImp::ProcessImpl(impl) => impl,

            #[cfg(feature = "mux_client")]
            RemoteChildImp::MuxProcessImpl(impl) => impl,
        }
    }
    fn to_trait_obj(&mut self) -> &dyn mut RemoteChildTrait {
        match self {
            RemoteChildImp::ProcessImpl(impl) => impl,

            #[cfg(feature = "mux_client")]
            RemoteChildImp::MuxProcessImpl(impl) => impl,
        }
    }
}
```

Then it can be used like this:

```
    stdin: (&mut imp).to_trait_obj().stdin().take().map(|val| val.into()),
```

Though it seems that your solution using macro is easier to implement.

Reply:
But if we are going to use the macro, then it might be better to put `Session` into a separate file...

So I will now try to the alternative way I come up with.

Reply:
> Needing these kinds of `match` blocks _everywhere_ is why I'm slightly partial to using
> 
> ```rust
> impl: Box<dyn RemoteChildImpl>
> ```
> 
> instead where `RemoteChildImpl` is a trait sweat_smile It's fine as-is too though, just food for thought if you'd be interested in trying it out.

The problem with `trait` isn't just using `Box` though, it is that every `async` function has to be converted to return a `Box<dyn Future<...>>`.

Reply:
I have simplified `Session`, `Command` and `RemoteChild` using the macro you gave me.